src/controller/DeviceUnpair.cpp

```diff
@@ -0,0 +1,187 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <controller/DeviceUnpair.h>
+
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/InteractionModelEngine.h>
+#include <controller-clusters/zap-generated/CHIPClusters.h>
+
+using namespace chip::app::Clusters;
+namespace chip {
+namespace Controller {
+
+DeviceUnpair::DeviceUnpair() :
+    mOnDeviceConnectedCallback(OnDeviceConnectedFn, this), mOnDeviceConnectionFailureCallback(OnDeviceConnectionFailureFn, this)
+{}
+
+void DeviceUnpair::RegisterCallback(Callback * callback)
+{
+    mCallback = callback;
+}
+void DeviceUnpair::UnpairDevice(DeviceProxy * proxy, NodeId remoteDeviceId)
+{
+    mRemoteDeviceId = remoteDeviceId;
+    PerformDeviceUnpairStep(proxy, UnpairDeviceStage::kReadCurrentFabricIndex, CHIP_NO_ERROR);
+}
+
+void DeviceUnpair::PerformDeviceUnpairStep(DeviceProxy * proxy, UnpairDeviceStage stage, CHIP_ERROR err)
+{
+    mProxy = proxy;
+    mStage = stage;
+
+    switch (stage)
+    {
+    case kReadCurrentFabricIndex: {
+        ReadCurrentFabricIndex(proxy);
+        break;
+    }
+    case kSendRemoveFabric: {
+        SendRemoveFabricIndex(proxy, mFabricIndex);
+        break;
+    }
+    case kUnpairError:
+    case kUnpairCleanUp:
+    default: {
+        FinishUnpairDevice(err);
+        break;
+    }
+    }
+}
+
+void DeviceUnpair::ReadCurrentFabricIndex(DeviceProxy * proxy)
+{
+    app::InteractionModelEngine * engine = app::InteractionModelEngine::GetInstance();
```

Comment:
Might be simpler to use OperationalCredentialsCluster here too.

Reply:
Thank you for your opionion. I updated this code.